src/ioctl/mod.rs

```diff
@@ -188,93 +184,66 @@ pub unsafe trait Ioctl {
     ) -> Result<Self::Output>;
 }
 
-/// The opcode used by an `Ioctl`.
-#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct Opcode {
-    /// The raw opcode.
-    raw: RawOpcode,
-}
-
-impl Opcode {
-    /// Create a new old `Opcode` from a raw opcode.
-    ///
-    /// Rather than being a composition of several attributes, old opcodes are
-    /// just numbers. In general most drivers follow stricter conventions, but
-    /// older drivers may still use this strategy.
-    #[inline]
-    pub const fn old(raw: RawOpcode) -> Self {
-        Self { raw }
-    }
+/// Const functions for computing opcode values.
+#[cfg(any(linux_kernel, bsd))]
+pub mod opcode {
+    use super::*;
 
     /// Create a new opcode from a direction, group, number, and size.
     ///
     /// This corresponds to the C macro `_IOC(direction, group, number, size)`
-    #[cfg(any(linux_kernel, bsd))]
     #[inline]
     pub const fn from_components(
         direction: Direction,
         group: u8,
         number: u8,
         data_size: usize,
-    ) -> Self {
-        assert!(
-            data_size <= RawOpcode::MAX as usize,
-            "data size is too large"
-        );
+    ) -> Opcode {
+        assert!(data_size <= Opcode::MAX as usize, "data size is too large");
 
-        Self::old(platform::compose_opcode(
+        platform::compose_opcode(
             direction,
-            group as RawOpcode,
-            number as RawOpcode,
-            data_size as RawOpcode,
-        ))
+            group as Opcode,
+            number as Opcode,
+            data_size as Opcode,
+        )
     }
 
     /// Create a new non-mutating opcode from a group, a number, and the type
     /// of data.
     ///
     /// This corresponds to the C macro `_IO(group, number)` when `T` is zero
     /// sized.
-    #[cfg(any(linux_kernel, bsd))]
     #[inline]
-    pub const fn none<T>(group: u8, number: u8) -> Self {
-        Self::from_components(Direction::None, group, number, mem::size_of::<T>())
+    pub const fn none<T>(group: u8, number: u8) -> Opcode {
```

Comment:
Given that the corresponding Linux `_IO` macro doesn't take an `argtype` like the others, I wonder if `T` could be removed here. I think the size could be hard-coded to 0.

Reply:
Makes sense, I've now added a patch for that.